### PR TITLE
Add Compact Commit List as an option

### PR DESCRIPTION
**Pull request by Gr3q**
Related feature request https://github.com/gitahead/gitahead/issues/261

**This feature is now usable with no side effects.**

Would be nice to have:
* I want to use the git like relative time string (26 minutes ago) as timestamp, and I don't know if I can pull that from git or I have to build it myself.
* I only want to display the date if it is different than the previous string, but I don't know how/where to store the previous value to make this work.

I excluded the author name on purpose, because it does not provide any reference or context (only if you have to look out for that One person's commits in your team...) and  takes up way too much space.

Current State:
image

### CODE REVIEW

File: src/ui/Badge.cpp [hunk 1]
```diff
@@ -80,11 +80,110 @@ QSize Badge::size(const QFont &font, const Label &label)
...
+{
+  if (alignment == RIGHT) {
+    return Badge::paintRight(painter, list, rect, opt, alignment);
+  }
+  else {
+    return Badge::paintLeft(painter, list, rect, opt, alignment);
+  }
+}
+
+int Badge::paintLeft(
```
Comment (Gr3q) [inline src/ui/Badge.cpp]: Lots of Duplicate code here

File: src/ui/Badge.cpp [hunk 2]
```diff
@@ -80,11 +80,110 @@ QSize Badge::size(const QFont &font, const Label &label)
...
+  BadgeAlignment alignment)
+{
+  if (alignment == RIGHT) {
+    return Badge::paintRight(painter, list, rect, opt, alignment);
+  }
+  else {
+    return Badge::paintLeft(painter, list, rect, opt, alignment);
+  }
+}
+
```
Comment (Gr3q) [inline src/ui/Badge.cpp]: Lots of duplicate code starting from here

File: src/ui/CommitList.cpp [hunk 3]
```diff
@@ -44,10 +44,11 @@
 #endif
 
 namespace {
+const bool compactMode = true;
```
Comment (Gr3q) [inline src/ui/CommitList.cpp]: Mode is dependent on this flag instead on the setting from the tickbox

File: src/ui/CommitList.cpp [hunk 4]
```diff
@@ -809,82 +814,160 @@ class CommitDelegate : public QStyledItemDelegate
...
+      QRect star = rect;
+
+      if (compactMode) {
+        int maxWidthRefs = (int)(rect.width() * 0.5); // Max 30% of the 
+        const int minWidthRefs = 50; // At least display The ellipsis
+        const int minWidthRequestDesc = 100;
+        int minDisplayWidthDate = 350;
+        int minDisplayWidthName = 800;
+        int minWidthName = 50;
+        int maxWidthName = (int)(rect.width() * 0.15);
```
Comment (Gr3q) [inline src/ui/CommitList.cpp]: All of these values control the behaviour of a single row, some of them could be factored to be able to set by the user (pretty unrealistic at the moment for me to do)

Anyway, they are all up for discussion, and I don't know how it behaves with different DPI and font sizes.

File: src/ui/CommitList.cpp [hunk 5]
```diff
@@ -809,82 +814,160 @@ class CommitDelegate : public QStyledItemDelegate
...
+
+        // Star always takes up its height on the right side
+        star.setX(star.x() + star.width() - star.height());
+        star.setY(star.y() - kVerticalMargin);
+
+        // maybe..maybe?? not needed in compact mode?
+        // can't really remove it, at least it 
+        // needs to be an option
```
Comment (Gr3q) [inline src/ui/CommitList.cpp]: I want to keep the ID visible, it is good reference, also it is also used in the context menu.

File: src/ui/CommitList.cpp [hunk 6]
```diff
@@ -809,82 +814,160 @@ class CommitDelegate : public QStyledItemDelegate
...
+        painter->drawText(box, Qt::AlignRight, id);
+        painter->restore();
+        box.setWidth(box.width() - idWidth);
+
+        // Draw date. Only if String is not the same as previous?
+        QDateTime date = commit.committer().date().toLocalTime();
+        QString timestamp =
+          (date.date() == QDate::currentDate()) ?
+          date.time().toString(Qt::DefaultLocaleShortDate) :
+          date.date().toString(Qt::DefaultLocaleShortDate);
```
Comment (Gr3q) [inline src/ui/CommitList.cpp]: I want to use the git like relative timestamp strings here

**Comment by dgalli1** [comment]
Would it be possible to do something like gitkraken for the authors? 
They generate a image for every user based on their git profile.
grafik

This doesn't take up much space. But helps a lot with looking up commits from a certain person. And it looks way slicker then initals. I don't think it has to necessary be implementet at the history as gitkraken is doing i think next to the timestamp would be a improvment aswell

File: src/ui/CommitList.cpp [hunk 7]
```diff
@@ -809,82 +814,160 @@ class CommitDelegate : public QStyledItemDelegate
...
+        if (box.width() > minWidthRequestDesc + fm.horizontalAdvance(timestamp) + 8 && totalWidth > minDisplayWidthDate) {
```
Comment (Gr3q) [inline src/ui/CommitList.cpp]: I intend to add a check with previous string != current string here, but I don't know where to store the previous string

**Comment by Gr3q** [comment]
> Would it be possible to do something like gitkraken for the authors?
> They generate a image for every user based on their git profile.
> grafik
> 
> This doesn't take up much space. But helps a lot with looking up commits from a certain person. And it looks way slicker then initals. I don't think it has to necessary be implementet at the history as gitkraken is doing i think next to the timestamp would be a improvment aswell


I believe it is possible to do this, although it is a design choice what's not for me to decide.

At the moment I'll try to focus to fix the issues with the existing features.

File: src/ui/CommitList.cpp (hunk 3 above)
Comment (Gr3q) [reply src/ui/CommitList.cpp]: Resolved

File: src/ui/CommitList.cpp (hunk 5 above)
Comment (Gr3q) [reply src/ui/CommitList.cpp]: kept in

**Comment by CFlix** [comment]
Thank you so much @Gr3q! This looks great and is an absoloutely necessary feature. Hope the PR gets merged soon to end up in the next release!

**Comment by Gr3q** [comment]
This PR is actually ready for merge, all the extra features can be implemented later (They turned out to be quite challenging and I do not have too much time to spend on this feature anymore)

If there is anything I should change/fix/additional testing please let me know.

Manual tests made in Linux:
Switching between Compact and Normal mode for an opened repo
Switching between Repos With different modes set
Opening new repo when an opened one was set to compact mode.

**Comment by kas-luthor** [comment]
The width of the short ref IDs doesn't seem to fit on Windows
Maybe a broader character should be used for measurements?

image

**Comment by Gr3q** [comment]
I wanted the commit reference width to be uniform (like a column), so the rest to its left are aligned correctly.

**Comment by Gr3q** [comment]
@kas-luthor Does the problem still persists? I will try to build it on Windows _(again, had problems with openssl before)_ as well today to check if the commit reference displays correctly.

**Comment by kas-luthor** [comment]
The width issue doesn't appear anymore

**Comment by hackhaslam** [comment]
It still doesn't quite compute the right text width for the id, at least on macOS:
Screen Shot 2019-11-18 at 7 03 50 PM

**Comment by Gr3q** [comment]
It seems I can't get away with simple hacks... Fine, I'll do it properly this time.

**Comment by kas-luthor** [comment]
Also: I think it would be nice to have an option to set compact mode as a default
Should I implement this or would you rather do it yourself?

**Comment by Gr3q** [comment]
@kas-luthor That's something that was on my list to do, you can do it if you have the time. Or if you tell me which section and group is the best place for this setting in Options I can do it too.

**Comment by kas-luthor** [comment]
I'd put it in the "Window" tab
I'll be able to implement it in ~6h when I get home

**Comment by Gr3q** [comment]
Ok, thank you.

**Comment by Gr3q** [comment]
Commit ID width calculation should be fixed now (it covers most cases). Widest characters are either 0,4,6,9,d or b, so I'm checking for those.

Optionally, I can use monospace font just for the commit ID to eliminate this problem, like this:

image

Although It can look weird if people use wildly different monospace font compared to the default. For example for me the font baseline is different.

**Comment by hackhaslam** [comment]
I'm reviewing this now.

**Comment by kas-luthor** [comment]
@hackhaslam 
My PR Gr3q/gitahead#2 is still open

**Comment by hackhaslam** [comment]
@kas-luthor I will look at it.

**Comment by Gr3q** [comment]
@kas-luthor thanks its merged now, I don't know why I don't get notifications about my fork.

**Comment by Gr3q** [comment]
@hackhaslam feel free to pick it apart, I think there are a few things what I implemented at the wrong places/incorrectly (probably in CommitList.cpp) what I'm happy to fix if you point them out.

**Comment by hackhaslam** [comment]
I tried it without duplicating the date in the way that I think you were trying to do. It looks weird to me. It would probably look better if the space were left blank instead of allowing the message to grow into that space. However, the width of the date string is variable, so I wouldn't really know how much space to leave. I'm going to merge it without this for now.

Screen Shot 2019-11-20 at 4 15 23 PM

**Comment by hackhaslam** [comment]
I merged this manually with a couple of additional commits. I also didn't use the default setting commit from @kas-luthor. Instead, I just changed the existing setting to an application global setting instead of a repo-specific one. I don't think there needs to be a separate repo-specific setting and global default setting, but I can revisit that if you guys think it's important.

@Gr3q @kas-luthor I'm going to start including contributor names for specific changes in the change log. Can you let me know your preferred name (real name or handle) if you want it appear?

Also, @kas-luthor I'm trying to review some of your pull requests now. I can implement the macOS-specific portions of the ones that need them.

**Comment by Gr3q** [comment]
I'm happy with the changes. My name is Attila Greguss, you can use it.

Thanks

**Comment by kas-luthor** [comment]
@hackhaslam 
Just "Kas" will be fine, thanks

**Pull request closed without merging**